java-symbolic-execution/src/main/java/org/sonar/java/se/checks/NullDereferenceCheck.java

```diff
@@ -120,12 +136,117 @@ public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
         return checkMemberSelect(context, (MemberSelectExpressionTree) syntaxNode, peekValue);
       case SYNCHRONIZED_STATEMENT:
         return checkConstraint(context, syntaxNode, peekValue);
+      case VARIABLE:
+        var variableTree = (VariableTree) syntaxNode;
+        extractUnboxingExpression(variableTree).ifPresent(expression ->
+          checkConstraint(context, expression, peekValue)
+        );
+        break;
       default:
-        // ignore
+        if (syntaxNode instanceof AssignmentExpressionTree) {
+          inspectAssignment(context, (AssignmentExpressionTree) syntaxNode, peekValue);
+        } else if (syntaxNode instanceof BinaryExpressionTree) {
+          inspectBinaryExpression(context, (BinaryExpressionTree) syntaxNode, peekValue);
+        } else if (syntaxNode instanceof UnaryExpressionTree) {
+          var unaryOperation = (UnaryExpressionTree) syntaxNode;
+          extractUnboxingExpression(unaryOperation).ifPresent(expression ->
+            checkConstraint(context, expression, peekValue)
+          );
+        }
```

Comment:
I am not sure if this might be out of scope for this PR, but we're not detecting this case:
```
  int illegalReturn() {
    Integer l = null;
    return l;
  }
```
We might be able to catch it by adding here the switch case and its relative logic for `Tree.Kind.RETURN_STATEMENT` .
WDYT?